### PR TITLE
feat(datasource/aws-machine-image): Add profile and region configuration support

### LINKED ISSUE #14421: Support for regions in AWS Resources (aws-machine-image)
### What would you like Renovate to be able to do?

In the meantime, Renovate does not support multiple regions. AWS has different resources in different region/location. From [the code ](https://github.com/renovatebot/renovate/blob/f5cc1de36f7d6f3ce7d689a9b6675a930d81337c/lib/datasource/aws-machine-image/index.ts#L38) the region is supplied with environment variable e.g. AWS_REGION or AWS_DEFAULT_REGION. First of all, this is not documented. Second, there is no way to switch regions on the fly. Filters do no support regions, its an external value.

1. [Docs](https://docs.renovatebot.com/modules/datasource/#aws-machine-image-datasource)

Quite common project structure, I did omit environments and other stuff

```
  ├── eu-west-1
  │       ├── values
  ├── eu-central-1
  │       ├── values
  ├── ap-southeast-1
  │       ├── values
```

Or it can be like that
```
  ├── somepath
  │       ├── values
  ├── someotherpath
  │       ├── values
```

In this case  values file, probably looks like that

```hcl
region="ap-southeast-1"
eks="1.19"
eks_ami=ami-11ce3d9008cab33cb
```

```hcl
region="eu-west-1"
eks="1.21"
eks_ami=ami-02ce3d9008cab69cb
```

Non of this cases is supported. As Renovate support a single region. There are probably workarounds.

### If you have any ideas on how this should be implemented, please tell us here.

The [code](https://github.com/renovatebot/renovate/blob/f5cc1de36f7d6f3ce7d689a9b6675a930d81337c/lib/datasource/aws-machine-image/index.ts#L38). Should support a region, it can be optional, so no current functionality is broken.

Documentation updated. As well as minimum AWS token scope should be provided as per [List Priviledged Access](https://docs.aws.amazon.com/IAM/latest/UserGuide/best-practices.html#grant-least-privilege)

Region can be read from file paths like below
```yml
environments/eu-west-1/dev
dev/eu-central-1/ec2
``` 

Or from file, if its already there
```hcl
aws_region="eu-west-1"
region="eu-west-1"
``` 

Or would recognize the following line, as from the docs
```
# With AMI name mentioned in the comments
# amiFilter=[{"Name":"owner-id","Values":["602401143452"]},{"Name":"name","Values":["amazon-eks-node-1.21-*"]}]
# region=eu-west-1
# currentImageName=unknown
my_ami1: ami-02ce3d9008cab69cb
```

`Edge case`
Some advanced teams do follow the Infrastructure [Manifest patterns](http://codyaray.com/2014/11/simplify-deployment-with-infrastructure-manifest)
When all the infrastructure configuration e.g. verssions in single file

```json
{
  "prod": {
    "us-east-1": {
      "appserve01ea1": {
        "applications": [
          "appserve"
        ],
        "zone": "us-east-1a",
        "fqdn": "ec2-1-2-3-4.compute-1.amazonaws.com",
        "private ip": "10.9.8.7",
        "public ip": "1.2.3.4",
        "id": "i-a1234bc5"
      },
      ...
  },
  ...
}
```

```yml
prod:
  - region: us-east-1
    id: i-a1234bc5
  - region: eu-west-1
    id: i-a1234bc5
dev:
  - region: us-east-1
    id: i-a123ccc5
```

### Is this a feature you are interested in implementing yourself?

Maybe

### Example repository

https://github.com/ik-workshop/renovate-issue-14421

### PR DESCRIPTION
**Pull request by reegnz**
The PR adds region support as well as profile support in case the user wants to utilize the shared config file to switch between configurations.

## Changes

The aws-machine-image datasource can optionally handed region and aws-profile configuration through the amiFilter json array.

## Context

This is an attempt at fixing #14421.
- fixes #14421

## Documentation (please check one with an [x])

- [X] I have updated the documentation, or
- [ ] No documentation update is required

## How I've tested my work (please select one)

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [x] Both unit tests + ran on a real repository

### CODE REVIEW

File: lib/modules/datasource/aws-machine-image/index.ts [hunk 1]
```diff
@@ -1,10 +1,19 @@
-import { DescribeImagesCommand, EC2Client, Image } from '@aws-sdk/client-ec2';
+import { fromNodeProviderChain } from '@aws-sdk/credential-providers';
+import { DescribeImagesCommand, Filter, EC2Client, Image } from '@aws-sdk/client-ec2';
 import { cache } from '../../../util/cache/package/decorator';
-import { Lazy } from '../../../util/lazy';
 import * as amazonMachineImageVersioning from '../../versioning/aws-machine-image';
 import { Datasource } from '../datasource';
 import type { GetReleasesConfig, ReleaseResult } from '../types';
 
+type AwsClientConfig = {
```
Comment (viceice) [inline lib/modules/datasource/aws-machine-image/index.ts:8]: use interface and move to types.ts

File: lib/modules/datasource/aws-machine-image/index.ts [hunk 2]
```diff
@@ -29,28 +38,57 @@ export class AwsMachineImageDataSource extends Datasource {
...
+
+  getAmiFilterCommand(filter: Filter[]): DescribeImagesCommand {
+    return new DescribeImagesCommand({
+      Filters: filter,
+    });
+  }
+
+  loadConfig(serializedAmiFilter: string): [Filter[], AwsClientConfig] {
+    const parsedConfig: ParsedConfig = JSON.parse(serializedAmiFilter)
+    return parsedConfig.reduce<[Filter[], AwsClientConfig]>((acc, elem) => {
```
Comment (viceice) [inline lib/modules/datasource/aws-machine-image/index.ts:69]: use for-of loop

**Review by viceice — COMMENTED** [review]
needs tests

**Comment by reegnz** [comment]
Not sure how to test for EC2 client configuration, the existing tests are all mocking the client.

**Comment by reegnz** [comment]
Going to write some jest tests to test the `loadConfig` function.

File: lib/modules/datasource/aws-machine-image/index.ts [hunk 3]
```diff
@@ -29,16 +35,47 @@ export class AwsMachineImageDataSource extends Datasource {
...
-
   private readonly now: number;
 
   constructor() {
     super(AwsMachineImageDataSource.id);
-    this.ec2 = new Lazy(() => new EC2Client({}));
     this.now = Date.now();
   }
 
+  isAmiFilter(config: Filter | AwsClientConfig): config is Filter {
```
Comment (viceice) [inline lib/modules/datasource/aws-machine-image/index.ts:45]: should be private

File: lib/modules/datasource/aws-machine-image/index.ts [hunk 4]
```diff
@@ -29,16 +35,47 @@ export class AwsMachineImageDataSource extends Datasource {
...
+    return 'Name' in config && 'Values' in config;
+  }
+
+  getEC2Client(config: AwsClientConfig): EC2Client {
```
Comment (viceice) [inline lib/modules/datasource/aws-machine-image/index.ts:49]: same

File: lib/modules/datasource/aws-machine-image/index.ts [hunk 5]
```diff
@@ -29,16 +35,47 @@ export class AwsMachineImageDataSource extends Datasource {
...
+    const { profile, region } = config;
+    return new EC2Client({
+      region,
+      credentials: profile ? fromNodeProviderChain({ profile }) : undefined,
+    });
+  }
+
+  getAmiFilterCommand(filter: Filter[]): DescribeImagesCommand {
```
Comment (viceice) [inline lib/modules/datasource/aws-machine-image/index.ts:57]: same

**Review by viceice — COMMENTED** [review]
pnpm lockfile is broken, needs fix

**Comment by reegnz** [comment]
That's very curious, might have broken it with some change. Pushed a fixed one.

**Review by viceice — COMMENTED** [review]
lint issues 😉

**Comment by reegnz** [comment]
ah, feel like such a noob! 😮‍💨 
Thanks for your patience. Will fix in a sec.

**Review by secustor — APPROVED** [review]

**Pull request merged**